index.js

```diff
@@ -14,52 +14,57 @@ export default function elementReady(selector, {
 	timeout = Number.POSITIVE_INFINITY,
 	predicate,
 } = {}) {
-	const cacheKeys = [selector, stopOnDomReady, timeout, waitForChildren, target];
-	const cachedPromise = cache.get(cacheKeys);
+	const cacheKey = [selector, stopOnDomReady, timeout, waitForChildren, target];
+	const cachedPromise = cache.get(cacheKey);
 	if (cachedPromise) {
 		return cachedPromise;
 	}
 
-	let rafId;
-	const deferred = pDefer();
-	const {promise} = deferred;
+	let shouldStop = false;
 
-	cache.set(cacheKeys, promise);
-
-	const stop = element => {
-		cancelAnimationFrame(rafId);
-		cache.delete(cacheKeys, promise);
-		deferred.resolve(element);
+	const stop = () => {
+		cache.delete(cacheKey, promise);
+		shouldStop = true;
 	};
 
 	if (timeout !== Number.POSITIVE_INFINITY) {
 		setTimeout(stop, timeout);
 	}
 
 	// Interval to keep checking for it to come into the DOM
-	(function check() {
-		const element = getMatchingElement({target, selector, predicate});
+	const promise = (async () => {
+		try {
+			for await (const _ of requestAnimationFrames()) { // eslint-disable-line no-unused-vars
+				if (shouldStop) {
+					return;
+				}
 
-		// When it's ready, only stop if requested or found
-		if (isDomReady(target) && (stopOnDomReady || element)) {
-			stop(element ?? undefined); // No `null`
-			return;
-		}
+				const element = getMatchingElement({target, selector, predicate});
 
-		let current = element;
-		while (current) {
-			if (!waitForChildren || current.nextSibling) {
-				stop(element);
-				return;
-			}
+				// When it's ready, only stop if requested or found
+				if (isDomReady(target) && (stopOnDomReady || element)) {
+					return element ?? undefined; // No `null`
+				}
 
-			current = current.parentElement;
-		}
+				let current = element;
+				while (current) {
+					if (!waitForChildren || current.nextSibling) {
+						return element;
+					}
 
-		rafId = requestAnimationFrame(check);
+					current = current.parentElement;
+				}
+			}
+		} finally {
+			cache.delete(cacheKey, promise);
+		}
 	})();
 
-	return Object.assign(promise, {stop: () => stop()});
+	promise.stop = stop;
```

Comment:
Not sure I understand? I just changed from using `Object.assign`.

Reply:
I didn't know it was a pre-existing API: https://github.com/sindresorhus/element-ready/blob/7917f35cabdbea87bf17c4a5c2f9dc0c6b2db96c/index.d.ts#L65-L72

I'd offer `signal` instead though

Reply:
They would behave differently though. `signal` throws, while this returns `undefined` on `stop()`. It may make sense, however, it's not something I plan to work on.

Reply:
> I didn't know it was a pre-existing API:

You added it 🤣

https://github.com/sindresorhus/element-ready/commit/32363fa9439cc432138aac754b62ba32532fc98c

Reply:
People change, cities are built, pandemics pass.